components/core/src/service.rs

```diff
@@ -150,6 +150,9 @@ impl ServiceGroup {
         where S1: AsRef<str>,
               S2: AsRef<str>
     {
+        if group.as_ref().contains('.') {
+            return Err(Error::InvalidGroupName(group.as_ref().to_string()));
+        }
```

Comment:
I'd rather we keep all our validation logic in a single place, and this special-cases one piece of it (and duplicates what's already encoded in the monster regex).

Incidentally, I notice that we have a `Self::validate` function that this is piped into two lines down, which should _also_ fail if the group has dot. 

I think we should actually remove `Self::validate`, since it is contractually identical to the `FromStr` trait (once we make sure that they're both doing the same thing, that is), and then just use `formatted.parse()` and be done with it.

This also means you could get rid of the new error case you added here (I want to overhaul our errors soon anyway, and the less we have the easier that becomes :smile: https://github.com/habitat-sh/habitat/issues/6557).

The more I look at this type, however, I'm suspicious that all the regex-based code is buying us is a bunch of confusion, and that a simple struct with four fields could serve us better. It may be worthwhile to see what that refactoring would look like in this PR, since we're touching validation, and that would simplify a lot of it. The big thing to look out for with that, however, is that we'd need to manually implement `Serialize` and `Deserialize`, as the derived implementation of the current `ServiceGroup` struct is very different from the ones derived on a struct with four fields. However, @davidMcneil has recently discovered some very useful code that we can take advantage of for that: see the thread [here](https://github.com/habitat-sh/habitat/pull/6539#discussion_r284346211) for further details.

Reply:
there are a couple nuances here that complicates things. There are several cli calls and internal APIs that deal with a "service group" as a single string. Which embeds (at the least) `<service>.<group>` and potentially environment and org as well. The regex parsing code in `Self::validate` and `FromStr` can parse all 4 elements. However I found the extra check on group was necessary because someone could pass `hab svc load origin/service my.group` and  you would end up passing the validation but have a service of `service.my` and a group of `group`.

since calls like `hab config apply` and `hab file upload` expect you to pass in a single "service group" string, it seems like you still need either a regex or some explicit parsing that does the same thing.

Reply:
Regardless of where we go from here (I am in favor of eliminating the regex in favor of a more direct, readable approach), the error we return would be more useful to users if it explains *why* the name is invalid.